connectors/byoc.py

```diff
@@ -668,71 +616,99 @@ async def prepare_docs(self, data_provider, pipeline=None, filtering=None):
 
-            if self.configuration.is_empty():
                 # sets the defaults and the flag to NEEDS_CONFIGURATION
-                self.doc_source[
+                self.configuration = doc[
                     "configuration"
                 ] = source_klass.get_simple_configuration()
-                self.doc_source["status"] = Status.NEEDS_CONFIGURATION.value
-                self._update_config(self.doc_source)
+                self.status = doc["status"] = Status.NEEDS_CONFIGURATION.value
                 logger.debug(f"Populated configuration for connector {self.id}")
-
-            # sync state if needed (when service type or configuration is updated)
-            try:
-                await self.sync_doc(force=False)
             except Exception as e:
                 logger.critical(e, exc_info=True)
-                raise ConnectorUpdateError(
-                    f"Could not update configuration for connector {self.id}"
+                raise DataSourceError(
+                    f"Could not instantiate {fqn} for {configured_service_type}"
                 )
+
+        try:
+            await self.index.update(doc_id=self.id, doc=doc)
         except Exception as e:
             logger.critical(e, exc_info=True)
-            raise DataSourceError(f"Could not instantiate {fqn} for {service_type}")
+            raise ConnectorUpdateError(
+                f"Could not update service type/configuration for connector {self.id}"
+            )
+
+    async def validate_filtering(self, validator):
+        draft_filter = self.filtering.get_draft_filter()
+        if not draft_filter.has_validation_state(FilteringValidationState.EDITED):
+            return
 
-        self.source_klass = source_klass
+        logger.info(f"Filtering of connector {self.id} is edited, validating...)")
+        validation_result = await validator.validate_filtering(draft_filter)
+        logger.info(f"The filtering is validated to be {validation_result.state.value}")
 
-    async def sync(self, sync_job_index, elastic_server, idling):
+        filtering = self.filtering.to_list()
+        for filter_ in filtering:
+            if filter_.get("domain", "") == Filtering.DEFAULT_DOMAIN:
+                filter_.get("draft", {"validation": {}})[
+                    "validation"
+                ] = validation_result.to_dict()
+                if validation_result.state == FilteringValidationState.VALID:
+                    filter_["active"] = filter_.get("draft")
+
+        self.filtering = filtering
+        await self.index.update(doc_id=self.id, doc={"filtering": filtering})
+
+    async def document_count(self):
+        await self.index.client.indices.refresh(
+            index=self.index_name, ignore_unavailable=True
+        )
+        result = await self.index.client.count(
+            index=self.index_name, ignore_unavailable=True
+        )
+        return result["count"]
+
+    async def sync(
```

Comment:
`sync` will eventually be removed once job execution is moved out of `Connector` class in a separate PR.

Part of this function (before job is claimed) will be moved to `SyncService`, and the rest (after job is created) will be moved to a new class `SyncJobRunner`.

Reply:
can you add these notes as `TODO` in the patch

Reply:
Done in https://github.com/elastic/connectors-python/pull/474/commits/1224df3fe0836512deffca22102671570b8ca9a7